test/unit/test_tpm2_policy.c

```diff
@@ -0,0 +1,374 @@
+// and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+//**********************************************************************;
+
+#include <errno.h>
+#include <stdarg.h>
+#include <stddef.h>
+
+#include <stdio.h>
+
+#include <setjmp.h>
+#include <cmocka.h>
+
+#include <sapi/tpm20.h>
+
+#include "pcr.h"
+#include "tpm2_alg_util.h"
+#include "tpm2_policy.h"
+#include "tpm2_session.h"
+#include "tpm2_util.h"
+
+typedef struct pcr_data pcr_data;
+struct pcr_data {
+    TPML_DIGEST values;
+};
+
+typedef struct test_file test_file;
+struct test_file {
+    char *path;
+    FILE *file;
+};
+
+/* Passing tests and static data are hardcoded around this sel spec */
+#define PCR_SEL_SPEC "sha256:0,1,2,3"
+
+/*
+ * Dummy value for the session handle read by teh wrapped version of:
+ *   Tss2_Sys_StartAuthSession
+ */
+#define SESSION_HANDLE 0xDEADBEEF
+
+/* dummy handle for sapi context */
+#define SAPI_CONTEXT   ((TSS2_SYS_CONTEXT *)0xDEADBEEF)
+
+/* Any PCR read returns this value */
+static TPM2B_DIGEST pcr_value = { .size = 32, .buffer = { 0x96, 0xa7, 0xfa,
+        0xaf, 0x16, 0x09, 0xb6, 0x50, 0xa4, 0xf2, 0x88, 0xc0, 0x90, 0x4f, 0x04,
+        0x83, 0x6e, 0xca, 0xda, 0x2f, 0x49, 0x78, 0x06, 0x94, 0x86, 0xa2, 0xbb,
+        0x02, 0xf2, 0xf0, 0x43, 0xea } };
+
+/* The expected hash for the pcr selection of sha256:0,1,2,3 */
+static TPM2B_DIGEST expected_policy_digest = { .size = 32, .buffer = { 0x62,
+        0x69, 0x69, 0xce, 0xa7, 0xc2, 0xea, 0x7a, 0xf8, 0x86, 0xf4, 0xb5, 0x09,
+        0xa5, 0xb8, 0x3a, 0xb3, 0x3b, 0x3a, 0x75, 0x75, 0x5d, 0x17, 0x40, 0x38,
+        0xa8, 0xd3, 0x33, 0x0f, 0xa7, 0x2a, 0xd4 } };
+
+TSS2_RC __wrap_Tss2_Sys_StartAuthSession(TSS2_SYS_CONTEXT *sysContext,
+        TPMI_DH_OBJECT tpmKey, TPMI_DH_ENTITY bind,
+        TSS2L_SYS_AUTH_COMMAND const *cmdAuthsArray,
+        const TPM2B_NONCE *nonceCaller,
+        const TPM2B_ENCRYPTED_SECRET *encryptedSalt, TPM2_SE sessionType,
+        const TPMT_SYM_DEF *symmetric, TPMI_ALG_HASH authHash,
+        TPMI_SH_AUTH_SESSION *sessionHandle, TPM2B_NONCE *nonceTPM,
+        TSS2L_SYS_AUTH_RESPONSE *rspAuthsArray) {
```

Comment:
and function definitions to something like this:

```
SS2_RC __wrap_Tss2_Sys_StartAuthSession(TSS2_SYS_CONTEXT *sysContext,
					TPMI_DH_OBJECT tpmKey, TPMI_DH_ENTITY bind,
					TSS2L_SYS_AUTH_COMMAND const *cmdAuthsArray,
					const TPM2B_NONCE *nonceCaller,
					const TPM2B_ENCRYPTED_SECRET *encryptedSalt,
					TPM2_SE sessionType,
					const TPMT_SYM_DEF *symmetric,
					TPMI_ALG_HASH authHash,
					TPMI_SH_AUTH_SESSION *sessionHandle,
					TPM2B_NONCE *nonceTPM,
					TSS2L_SYS_AUTH_RESPONSE *rspAuthsArray) {
```

Reply:
I don't like the large leading whitespace, as now it makes things go way horizontal for no benefit. In fact, your example is being cut off by Github.

Reply:
Things like TSS2...Create() would also benefit from the builder pattern like for TSS2...startauthsession()

Reply:
> I don't like the large leading whitespace, as now it makes things go way horizontal for no benefit. In fact, your example is being cut off by Github.

Ok, I disagree on the benefit since I found more readable but I don't have a strong opinion either.

Reply:
@martinezjavier Im fine with the argument list stacked vertically, just with less indentation, ie this:
```
SS2_RC __wrap_Tss2_Sys_StartAuthSession(TSS2_SYS_CONTEXT *sysContext,
        TPMI_DH_OBJECT tpmKey, TPMI_DH_ENTITY bind,
        TSS2L_SYS_AUTH_COMMAND const *cmdAuthsArray,
        const TPM2B_NONCE *nonceCaller,
        const TPM2B_ENCRYPTED_SECRET *encryptedSalt,
        TPM2_SE sessionType,
        const TPMT_SYM_DEF *symmetric,
        TPMI_ALG_HASH authHash,
        TPMI_SH_AUTH_SESSION *sessionHandle,
        TPM2B_NONCE *nonceTPM,
        TSS2L_SYS_AUTH_RESPONSE *rspAuthsArray) {
```